chef/cookbooks/utils/libraries/restart_manager.rb

```diff
@@ -0,0 +1,52 @@
+module ServiceRestart
+  class RestartManager
+    def initialize(cookbook_name, node)
+      @cookbook_name = cookbook_name
+      @node = node
+    end
+
+    def cookbook
+      # if the cookbook_name is apache2, then any cookbook could have triggered it, so
+      # try to pick the value from the extra key that we set
+      if @cookbook_name == "apache2"
+        @node.run_state["apache2_restart_origin"]
+      else
+        @cookbook_name.to_s
+      end
+    end
+
+    def add_restart_management_node_attributes
+      @node.set[:crowbar_wall] = {} unless @node[:crowbar_wall]
+      @node.set[:crowbar_wall][:requires_restart] = {} \
+            unless @node[:crowbar_wall][:requires_restart]
+      @node.set[:crowbar_wall][:requires_restart][cookbook] = {} \
+            unless @node[:crowbar_wall][:requires_restart][cookbook]
+    end
+
+    def disallow_restart?
+      # if the databag or item does not exits it returns a 404
+      data_bag = data_bag_item("crowbar-config", "disallow_restart") rescue {}
+
+      data_bag[cookbook] || false
+    end
+
+    def register_restart_request(service_name, is_pacemaker_service)
+      # Store more data about the service attempted restart
+      # use the service as the key name for the data for easy removal
+      # force it to be a simple hash instead of a node attribute if you want the
+      # update method to actually update. If you update the attribute directly it doesn't
+      # update but overwrites! The joys of chef!
```

Comment:
Can we call `add_restart_management_node_attributes` from here, instead of having the users of this lib do it? The rationale is that we would only create `@node[:crowbar_wall][:requires_restart]` if really required. And it matters because in the rails app, we do a search query for nodes with that attribute -- right now, all nodes would have that attribute, which is not great.

Reply:
Umm that is a good point and indeed, we are setting the attributes in every node for no reason.

Changing it!